tensorflow_probability/python/experimental/mcmc/sample_fold.py

```diff
@@ -0,0 +1,118 @@
+from tensorflow.python.util import nest
+
+
+__all__ = [
+    'sample_fold',
+]
+
+
+def sample_fold(
+    num_steps,
+    current_state,
+    previous_kernel_results=None,
+    kernel=None,
+    reducers=None,
+    num_burnin_steps=0,
+    num_steps_between_results=0,
+    parallel_iterations=10,
+    seed=None,
+    name=None,
+):
+  """Finalizes `num_steps` Reducer steps with `kernel` samples.
+
+  This driver will appropriately apply `WithReductions` and
+  `SampleDiscardingKernel` over the given `kernel` and `reducers`. It will
+  then update reducer states `num_steps` times (perform `num_steps` calls to
+  the `Reducer`'s `one_step` method). An arbitrary collection of `reducers`
+  can be provided, and the resulting finalized statistic(s) will be returned
+  in an identical structure.
+
+  Args:
+    num_steps: Integer number of `Reducer` steps.
+    current_state: `Tensor` or Python `list` of `Tensor`s representing the
+      current state(s) of the Markov chain(s).
+    previous_kernel_results: A `Tensor` or a nested collection of `Tensor`s.
+      Warm-start for the auxiliary state needed by the given `kernel`.
+      If not supplied, `sample_fold` will cold-start with
+      `kernel.bootstrap_results`.
+    kernel: An instance of `tfp.mcmc.TransitionKernel` which implements one step
+      of the Markov chain.
+    reducers: A (possibly nested) structure of `Reducer`s to be evaluated
+      on the `kernel`'s samples. If no reducers are given (`reducers=None`), an
+      empty list will be returned in place of streaming calculations.
+    num_burnin_steps: Integer number of chain steps to take before starting to
+      collect results. Defaults to 0 (i.e., no burn-in).
+    num_steps_between_results: Integer number of chain steps between collecting
+      a result. Only one out of every `num_steps_between_samples + 1` steps is
+      included in the returned results. Defaults to 0 (i.e., no thinning).
+    parallel_iterations: The number of iterations allowed to run in parallel. It
+      must be a positive integer. See `tf.while_loop` for more details.
+    seed: Optional seed for reproducible sampling.
+    name: Python `str` name prefixed to Ops created by this function.
+      Default value: `None` (i.e., 'mcmc_sample_fold').
+
+  Returns:
+    reduction_results: A (possibly nested) structure of finalized reducer
+      statistics. The structure identically mimics that of `reducers`
+    warm_restart_package: `Tuple` with the final state of the Markov chain(s)
+      and kernel results, including those of applied wrapper kernels (i.e.
+      `WithReductions`).
+  """
+  with tf.name_scope(name or 'mcmc_sample_fold'):
+    num_steps = tf.convert_to_tensor(
+        num_steps, dtype=tf.int32, name='num_steps')
+    current_state = tf.nest.map_structure(
+        lambda x: tf.convert_to_tensor(x, name='current_state'),
+        current_state)
+    if reducers is None:
+      reducers = []
```

Comment:
Is this actually needed?  Does WithReductions fail to do the right thing if the reducers argument is `None`?

Reply:
It was since `WithReductions` didn't have a None check (oops). I will add that to the next commit though

Reply:
WithReductions isn't currently implemented to take in a `None` reducer (it'll break). After all, with no reducer, why even use WithReductions?

One could have WithReductions return None as the streaming calculations field if the reducers are None, but then `sample_fold`'s call to `step_kernel` invokes a `tf.while` via `smart_for_loop` that's not too happy with a None value in the kernel results (can't be converted to a `Tensor`). If we really wanted to return `None`, we could make reducers an empty list and then just do some post-processing to make everything relevant being returned into `None`, but then the kernel results also aren't compatible with a `tf.while_loop` anymore (probably can't go through future drivers for warm restart), which seems problematic.

I've gone ahead and made this fix anyways as in situations like these in the past, I've always overlooked something haha